src/kernel/thread.c

```diff
@@ -616,6 +616,41 @@ void endTimeslice(bool_t can_timeout_fault)
         postpone(NODE_STATE(ksCurSC));
     }
 }
+
+bool_t ensureSchedulable(tcb_t *thread, bool_t can_timeout)
```

Comment:
this function is really big and branchy and not inlined and then called in a lot of places.

If possible it would be a lot better to create a set of specific functions that handle the exact cases we are dealing with inline rather than using a generic, branch predictor smashing function.

Reply:
There is only one case (the SC unbind) where we don't need to perform all of these checks together. In the SC unbind case there is always no SC. In every other case we have no static guarantee on whether or not the thread has an SC or whether that SC is configured or ready.

Reply:
I can probably move this function into a header and make it inline. The common case should be that all the top level checks are true.

Reply:
- in SC bind you also know that  the scheduling context is set.
- you should write the function to be fastest in the good case

something like

```fault_type_t faultType = Fault_Null_Fault
    if (likely(sc != NULL && sc->scRefillMax == 0 && refill_ready(sc)) {
        return true;
    } 

    bool valid = validTimeoutHandler(thread)
    word_t badge = 0
    if (sc == NULL) {
        faultType = seL4_Timeout_NoSC;
    } else if (sc->scRefillMax == 0) {
        faultType = seL4_Timeout_Unconfigured
        badge = sc->scBadge
    } else { /* out of time */
       faultType =  seL4_Timeout_Exhausted;
       badge = sc->scBadge
    }

    if (canTimeout && validTimeoutHandler) {
        current_fault = seL4_Fault_Timeout_new(badge, faultType);
        handleTimeout(fault);
    } else if (faultType == seL4_TimeoutExhausted) {
        postpone(thread);
    }
    return false;
}'''

Reply:
I'm about to push a slightly different approach that should inline a bit better.

Reply:
Alright, you might want to check over this commit again. The `can_timeout` argument is removed as the time exhaustion applies to all non-round-robin SCs under all circumstances (so a check is added for round-robin on the SC).

The outermost checks form the inlined function but are marked unlikely so as to have the same effect as the single likely condition you have above (i.e. push the body of each unlikely branch out of where the branch predictor will go) and the inner check is consolidated into a single function which gets called via an indirect branch.